src/libstd/keyword_docs.rs

```diff
@@ -1578,12 +1578,144 @@ mod type_keyword {}
 
 #[doc(keyword = "unsafe")]
 //
-/// Code or interfaces whose [memory safety] cannot be verified by the type system.
+/// Code or interfaces whose [memory safety] cannot be verified by the type
+/// system.
 ///
-/// The documentation for this keyword is [not yet complete]. Pull requests welcome!
+/// The `unsafe` keyword has two uses: to declare the existence of contracts the
+/// compiler can't check, and to declare that a programmer has checked that
+/// these contracts have been upheld. They are not mutually exclusive, as can be
+/// seen in `unsafe fn` (more details on this can be found below).
+///
+/// Unsafe Rust add some capabilities to Safe Rust:
+///
+/// - Dereference [raw pointers]
+/// - Call `unsafe` functions (including foreign functions, some compiler
+/// intrinsics and the raw allocator)
+/// - Implement `unsafe` [`trait`]s
+/// - Mutate [`static`]s (including [`extern`]al ones)
+/// - Access fields of [`union`]s
+///
+/// # The different meanings of `unsafe`
+///
+/// Depending on the place where it is used, the `unsafe` keyword meaning can
+/// change. See the following [discussion on Rust Internals] for more in-depth
+/// explanations.
+///
+/// Summarizing the main points:
+///
+/// - `unsafe fn`: calling this function means abiding by a contract the
+/// compiler cannot enforce.
+/// - `unsafe trait Trait`: implementing this [`trait`] means abiding by a
+/// contract the compiler cannot enforce.
+/// - `unsafe {}`: the invariants necessary to call the operations inside the
+/// block have been checked and are guaranteed to be valid.
+/// - `unsafe impl`: the invariants necessary to implement the trait have been
+/// checked and are guaranteed to be valid.
+///
+/// As of now (Rust 1.44), `unsafe fn` can act like an `unsafe` block around
+/// the code inside the function or a signal to callers that some conditions
+/// must be met before using it, or both at the same time.
+///
+/// Mixing these two meanings can be confusing and [discussion]s exist to use
```

Comment:
I think "proposal" would be a better term here than "discussion".

Reply:
I'll update it.